File: aoc2023/day2/csharp/Program.cs
// See https://aka.ms/new-console-template for more information

public class Program
{
    public static void Main(string[] args)
    {
        int r1 = 0;
        int r2 = 0;
        while (Console.ReadLine() is { } line)
        {
            ReadOnlySpan<char> l = line.AsSpan();
            var idx = l.IndexOf(':');
            var split = l[(idx + 1)..].Split([';', ',']);
            int red = 0;
            int blue = 0;
            int green = 0;
            while (split.TryNext(out var info))
            {
                info = info.Trim();
                var i = info.IndexOf(' ');
                int val = int.Parse(info[..i]);

                (red, blue, green) = info[(i + 1)..] switch
                {
                    "red" => (Math.Max(red, val), blue, green),
                    "green" => (red, blue, Math.Max(green, val)),
                    "blue" => (red, Math.Max(blue, val), green),
                    _ => throw new Exception("Unknown color")
                };
            }

            r2 += red * blue * green;
            if (red <= 12 && blue <= 14 && green <= 13)
            {
                // parse id
                var gameInfo = l[..idx];
                idx = gameInfo.IndexOf(' ');
                gameInfo = gameInfo[(idx + 1)..];
                var id = int.Parse(gameInfo);
                r1 += id;
            }
        }

        Console.WriteLine($"P1: {r1}; P2: {r2}");
    }
}

public ref struct Split<T>
where T : IEquatable<T>
{
    private ReadOnlySpan<T> inner;
    private T[] del;

    public Split(T del, ReadOnlySpan<T> toSplit)
    {
        this.del = [del];
        this.inner = toSplit;
    }

    public Split(T[] del, ReadOnlySpan<T> toSplit)
    {
        this.del = del;
        this.inner = toSplit;
    }

    public bool TryNext(out ReadOnlySpan<T> result)
    {
        if (this.inner.IsEmpty)
        {
            result = ReadOnlySpan<T>.Empty;
            return false;
        }

        var idx = this.inner.IndexOfAny(del.AsSpan());
        if (idx == -1)
        {
            result = this.inner;
            this.inner = ReadOnlySpan<T>.Empty;
        }
        else
        {
            result = this.inner[..idx];
            this.inner = this.inner[(idx + 1)..];
        }

        return true;
    }
}

file static class Extension
{
    public static Split<T> Split<T>(this ReadOnlySpan<T> self, T del)
        where T : IEquatable<T>
        => new Split<T>(del, self);

    public static Split<T> Split<T>(this ReadOnlySpan<T> self, T[] del)
        where T : IEquatable<T>
        => new Split<T>(del, self);
}

File: aoc2023/day2/fsharp/Program.fs
open System

exception UnknownColor of string

module P1 =
    let processLine(l:string) =
        // split game info and drawing info
        let parts = l.Split(':', 2)
        // get id
        let id = Array.item 1 (parts[0].Split ' ') |> int
        let mutable red, green, blue = (0,0,0)
        for info in parts[1].Split [| ','; ';'|] do
            let infoParts = info.Trim().Split ' '
            let count = infoParts[0] |> int
            match infoParts[1] with
            | "red" -> red <- Math.Max(red, count)
            | "blue" -> blue <- Math.Max(blue, count)
            | "green" -> green <- Math.Max(green, count)
            | _ -> raise (UnknownColor("Unknown color"))
        let power = red * blue * green
        if red <= 12 && blue <= 14 && green <= 13 then (id, power) else (0, power)

    let p1 : int * int =
        let mutable p1Result, p2Result = (0,0)
        let mutable eos = false
        while not eos do
            let line = Console.ReadLine()
            if line = null then
                eos <- true
                ()
            else
                let id, power = processLine line
                p1Result <- p1Result + id
                p2Result <- p2Result + power

        (p1Result, p2Result)
        


// For more information see https://aka.ms/fsharp-console-apps
[<EntryPoint>]
let main args =
    match P1.p1 with
    | (r1, r2) -> printfn "returning result: P1 - %d; P2 - %d" r1 r2
    0
